projects/client/RabbitMQ.Client/src/client/impl/AutorecoveringConnection.cs

```diff
@@ -1044,9 +1057,11 @@ private async Task RecoveryLoopRecoveringHandler(RecoveryCommand command)
                     }
                     else
                     {
-                        await Task.Delay(_factory.NetworkRecoveryInterval);
-                        _recoveryLoopCommandQueue.Enqueue(RecoveryCommand.PerformAutomaticRecovery);
-                        _semaphore.Release();
+                        Task.Delay(_factory.NetworkRecoveryInterval)
+                            .ContinueWith(t =>
+                                {
+                                    _recoveryLoopCommandQueue.Enqueue(RecoveryCommand.PerformAutomaticRecovery);
+                                });
```

Comment:
This is not guaranteed to run, since the Task returned form the Task.Delay + ContinueWith is never assigned to a variable. This will just be kicked off to a background thread, and might just end up being killed without finishing since noone is hanging on the the resulting task.

Reply:
@stebet Based on what are you making these claims that

- Task.Delay will not be run
- Will be "magically killed"

if never heard such a behavior and would love to profit from your experience

Side note: The continuation also doesn't use the scheduler hints to for example allow synchronous execution to not reschedule the continuation which can unnecessarily waste ressources. In general unless really on the hot path continuation style should be avoided and async/await should be preferred because it is less error prone, the statemachine takes care of all the special cases a continuation style would need to handle and in general the statemachine allocations are getting more and more optimized away in NETCore so the benefit of continuation style does rarely ever outweigh the risk, see also

https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-await-over-continuewith

Reply:
@danielmarbach Yes, you are correct :) My mind was stuck in pre 4.5 land when an unobserved task that might throw an exception caused finalizers to go haywire: https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/file-schema/runtime/throwunobservedtaskexceptions-element?redirectedfrom=MSDN#remarks

I'll amend my comments. Thanks for pointing it out :)

Reply:
A quick note though, that without awaiting the Task.Delay and the resulting ContinueWith, the method execution will certainly return before the task has finished running. If that was intended to run in the background, that's great, but maybe adding a comment for that purpose would be good to clarify.

Reply:
I updated the code to make my intention clearer. However, I am open to following @stebet 's suggestion below to revert the handlers back to async methods with cancellationtokens. Alternatively I'd be interested if there was a better (non-continuation) way you all knew of to schedule a signal from a thread pool thread without blocking or awaiting.